### PR TITLE
Adds Cucumber Tests for Email functionality

### PR DESCRIPTION
**Pull request by seb-siem-softwire**
### Context

As part of having a comprehensive test suite of our app, we need cucumber tests for our emails with Cypress

### Changes proposed in this pull request

Adds Cucumber tests for email

### Testing

Using Cucumber

### CODE REVIEW

**Comment by seb-siem-softwire** [comment]
Problems with mimemagic prevents runing the linting stage - other than that its ready for review

File: spec/cypress/integration/admin/AdminManagement.feature [hunk 1]
```diff
@@ -3,7 +3,7 @@ Feature: Admin user modifying admin users
 
   Background:
     Given I am logged in as an "admin"
-    And scenario "admin/administrators/manage_admin_users" has been run
+    And Admin account was created with email "emma-dow@example.com" and name "Emma Dow"
```
Comment (callumacrae) [inline spec/cypress/integration/admin/AdminManagement.feature:6]: We've been using scenario files for this - I think it would be better to continue that pattern

Comment (callumacrae) [inline spec/cypress/integration/admin/AdminManagement.feature:6]: What's the reason for this change?

Comment (seb-siem-softwire) [reply spec/cypress/integration/admin/AdminManagement.feature:6]: Better readability - we can replace that with more descriptive text

Comment (seb-siem-softwire) [reply spec/cypress/integration/admin/AdminManagement.feature:6]: It achieves the same thing - it uses factory bot to create the record

File: spec/cypress/integration/notifications/AdminAccountCreatedEmail.feature [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    Then I should be on "admin creation" page
+
+    When I type "Joe Wick" into "name input"
+    And I type "new-admin@example.com" into "email input"
+    And I click the submit button
+    And "main" should contain "Joe Wick"
+    And "main" should contain "new-admin@example.com"
+    And I click the submit button
+    Then "notification banner" should contain "User added"
+    And An Admin account created email will be sent to the email "new-admin@example.com"
```
Comment (callumacrae) [inline spec/cypress/integration/notifications/AdminAccountCreatedEmail.feature:17]: What's the reason this has been added as a new test instead of being added to the existing test?

File: spec/cypress/integration/notifications/SignInEmail.feature [hunk 3]
```diff
@@ -0,0 +1,26 @@
+Feature: Sign In Email
```
Comment (callumacrae) [inline spec/cypress/integration/notifications/SignInEmail.feature:1]: In my opinion, sign in email isn't a feature - sign in is, and email is part of that

File: spec/cypress/support/commands.js [hunk 4]
```diff
@@ -72,6 +74,50 @@ Cypress.Commands.add("verifySignInEmailSent", (user) => {
   });
 });
 
+Cypress.Commands.add("verifySignInEmailSentForEmail", (email) => {
```
Comment (callumacrae) [inline spec/cypress/support/commands.js:77]: I'm not sure about having these in the commands file as they're only being used in one place

File: spec/cypress/integration/notifications/SignInEmail.feature (hunk 3 above)
Comment (seb-siem-softwire) [reply spec/cypress/integration/notifications/SignInEmail.feature:1]: Cool - I can change the name

File: spec/cypress/integration/notifications/AdminAccountCreatedEmail.feature (hunk 2 above)
Comment (seb-siem-softwire) [reply spec/cypress/integration/notifications/AdminAccountCreatedEmail.feature:17]: I can move it to Admin Management feature and add it as last step - main reason was to have explicit Notification tests that tell a story from the perspective of sending notifications. I guess the key question is how to achieve a balance - should we have separate features each telling about different aspect or should we focus on long features that tests everything inbetween.

File: spec/cypress/support/commands.js (hunk 4 above)
Comment (seb-siem-softwire) [reply spec/cypress/support/commands.js:77]: Would you move those commands into another file? I do like having cypress commands - they do make the code better

File: spec/cypress/integration/notifications/AdminAccountCreatedEmail.feature (hunk 2 above)
Comment (callumacrae) [reply spec/cypress/integration/notifications/AdminAccountCreatedEmail.feature:17]: We already have multi-domain tests as writing specific tests for everything would slow the tests down a lot

This test takes three seconds to run, according to the github action

File: spec/cypress/support/commands.js (hunk 4 above)
Comment (callumacrae) [reply spec/cypress/support/commands.js:77]: I'd probably put the code directly in the step definition, I'm not sure what the advantage of splitting it across multiple files is when it isn't being reused!

File: spec/cypress/support/step_definitions/notifications.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    cy.verifyAdminAccountCreatedEmailSentForEmail(email);
+  }
+);
+
+Given(
+  "Email should be sent to Primary Email Contact of the School belonging to {string}",
+  (email) => {
+    cy.verifyPrimaryContactEmailSentForEmail(email);
+  }
+);
```
Comment (kicferk1) [inline spec/cypress/support/step_definitions/notifications.js:22]: "will be sent", "should be sent", ...

I would standardise to say "xxx email was/is/should be/will be sent to {string}". And stick with the sentence type. Out of them, I would go for 'should be sent'.

**Review by kicferk1 — COMMENTED** [review]
Mostly grammar nitpicks, I think it looks pretty good.

**Comment by kicferk1** [comment]
Given how many grammar debates we have over tests, do we want to pull some non-technical person in them?

File: spec/cypress/support/step_definitions/notifications.js (hunk 5 above)
Comment (callumacrae) [reply spec/cypress/support/step_definitions/notifications.js:22]: Yep, we've been using "should" a lot already

**Comment by callumacrae** [comment]
> Given how many grammar debates we have over tests, do we want to pull some non-technical person in them?

I think we should probably stop having so many grammar debates - they're really not the important part of the tests 😆

File: spec/cypress/integration/notifications/AdminAccountCreatedEmail.feature (hunk 2 above)
Comment (seb-siem-softwire) [reply spec/cypress/integration/notifications/AdminAccountCreatedEmail.feature]: Moved to Account Management

File: spec/cypress/integration/notifications/SignInEmail.feature (hunk 3 above)
Comment (seb-siem-softwire) [reply spec/cypress/integration/notifications/SignInEmail.feature]: Done

File: spec/cypress/support/commands.js (hunk 4 above)
Comment (seb-siem-softwire) [reply spec/cypress/support/commands.js:77]: The problem with moving the code around was that it depends in one of the functions in commands.js for computing email headers. Also other commands in commands.js also are being used in one place. How would you address this in terms of dependency on computeHeadersFromEmail function.

[the remark this answers is not in the document]
Comment (seb-siem-softwire) [reply spec/cypress/support/commands.js:77]: Done - changed the names

File: spec/cypress/support/step_definitions/accounts.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import { Given } from "cypress-cucumber-preprocessor/steps";
+
+Given("Admin account was created with {string}", (emailString) => {
```
[the remark this answers is not in the document]
Comment (seb-siem-softwire) [reply spec/cypress/support/step_definitions/accounts.js]: Done

File: spec/cypress/support/step_definitions/accounts.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        "create",
+        "user",
+        "admin",
+        { email: emailString, full_name: nameString },
+      ],
+    ]).as("userData");
+  }
+);
+
+Given("Lead Provider account was created with {string}", (emailString) => {
```
[the remark this answers is not in the document]
Comment (seb-siem-softwire) [reply spec/cypress/support/step_definitions/accounts.js]: Changed the names

File: spec/cypress/support/step_definitions/notifications.js (hunk 5 above)
Comment (seb-siem-softwire) [reply spec/cypress/support/step_definitions/notifications.js]: Changed to should

File: spec/cypress/support/step_definitions/notifications.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { Given, When } from "cypress-cucumber-preprocessor/steps";
+
+Given(
+  "An email sign in notification should be sent for email {string}",
```
[the remark this answers is not in the document]
Comment (seb-siem-softwire) [reply spec/cypress/support/step_definitions/notifications.js]: changed to "for the email"

File: spec/cypress/support/commands.js (hunk 4 above)
Comment (callumacrae) [reply spec/cypress/support/commands.js:77]: The step definitions are also defined in a single file :)

Comment (seb-siem-softwire) [reply spec/cypress/support/commands.js]: Done

**Review by callumacrae — APPROVED** [review]
